File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/service/FiatService.java
package com.taurushq.sdk.protect.client.service;

import com.google.common.base.Strings;
import com.taurushq.sdk.protect.client.mapper.ApiExceptionMapper;
import com.taurushq.sdk.protect.client.mapper.FiatMapper;
import com.taurushq.sdk.protect.client.model.ApiException;
import com.taurushq.sdk.protect.client.model.ApiRequestCursor;
import com.taurushq.sdk.protect.client.model.FiatProvider;
import com.taurushq.sdk.protect.client.model.FiatProviderAccount;
import com.taurushq.sdk.protect.client.model.FiatProviderAccountResult;
import com.taurushq.sdk.protect.client.model.FiatProviderCounterpartyAccount;
import com.taurushq.sdk.protect.client.model.FiatProviderCounterpartyAccountResult;
import com.taurushq.sdk.protect.client.model.FiatProviderEntityResult;
import com.taurushq.sdk.protect.client.model.FiatProviderOperation;
import com.taurushq.sdk.protect.client.model.FiatProviderOperationResult;
import com.taurushq.sdk.protect.client.model.Pagination;
import com.taurushq.sdk.protect.openapi.ApiClient;
import com.taurushq.sdk.protect.openapi.api.FiatApi;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProviderAccountReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProviderAccountsReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProviderCounterpartyAccountReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProviderCounterpartyAccountsReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProviderEntitiesReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProviderOperationReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProviderOperationsReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetFiatProvidersReply;

import java.util.Collections;
import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Service for managing fiat currency operations in the Taurus Protect system.
 * <p>
 * This service provides access to fiat provider accounts, counterparty accounts,
 * and operations for fiat currency management.
 * <p>
 * Example usage:
 * <pre>{@code
 * // List all fiat providers
 * List<FiatProvider> providers = client.getFiatService().getFiatProviders();
 *
 * // Get a specific account
 * FiatProviderAccount account = client.getFiatService()
 *     .getFiatProviderAccount("account-123");
 *
 * // List a provider's accounts, first page
 * FiatProviderAccountResult result = client.getFiatService()
 *     .getFiatProviderAccounts("provider", "label", null, null, 20, null);
 * // next page: pass result.getPage().getNextCursor() while result.getPage().hasMore()
 * }</pre>
 *
 * @see FiatProvider
 * @see FiatProviderAccount
 */
public class FiatService {

    /**
     * The underlying OpenAPI client for fiat operations.
     */
    private final FiatApi fiatApi;

    /**
     * Mapper for converting OpenAPI exceptions to SDK exceptions.
     */
    private final ApiExceptionMapper apiExceptionMapper;

    /**
     * Mapper for converting fiat DTOs to domain models.
     */
    private final FiatMapper mapper;

    /**
     * Instantiates a new Fiat service.
     *
     * @param openApiClient      the OpenAPI client
     * @param apiExceptionMapper the API exception mapper
     */
    public FiatService(final ApiClient openApiClient, final ApiExceptionMapper apiExceptionMapper) {
        checkNotNull(openApiClient, "openApiClient cannot be null");
        checkNotNull(apiExceptionMapper, "apiExceptionMapper cannot be null");

        this.apiExceptionMapper = apiExceptionMapper;
        this.fiatApi = new FiatApi(openApiClient);
        this.mapper = FiatMapper.INSTANCE;
    }

    /**
     * Retrieves all configured fiat providers.
     *
     * @return the list of fiat providers
     * @throws ApiException if the API call fails
     */
    public List<FiatProvider> getFiatProviders() throws ApiException {
        try {
            TgvalidatordGetFiatProvidersReply reply = fiatApi.fiatProviderServiceGetFiatProviders();
            return mapper.fromProviderDTOList(reply.getFiatProviders());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a fiat provider account by ID.
     *
     * @param id the account ID
     * @return the fiat provider account
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if id is null or empty
     */
    public FiatProviderAccount getFiatProviderAccount(final String id) throws ApiException {
        checkArgument(!Strings.isNullOrEmpty(id), "id cannot be null or empty");

        try {
            TgvalidatordGetFiatProviderAccountReply reply = fiatApi.fiatProviderServiceGetFiatProviderAccount(id);
            return mapper.fromAccountDTO(reply.getResult());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a page of a provider's accounts.
     *
     * @param provider    the provider, required
     * @param label       the provider configuration label, required
     * @param accountType filter by account type (optional)
     * @param sortOrder   sort order for results (optional, "ASC" or "DESC")
     * @param pageSize    the page size, null or 0 for the default
     * @param cursor      a previous page's {@code getPage().getNextCursor()}, null for the first page
     * @return the accounts and their page
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if provider or label is empty, or the page size is out of range
     */
    public FiatProviderAccountResult getFiatProviderAccounts(final String provider, final String label,
                                                              final String accountType, final String sortOrder,
                                                              final Integer pageSize, final String cursor)
            throws ApiException {
        return getFiatProviderAccounts(provider, label, accountType, sortOrder, Pagination.page(pageSize, cursor));
    }

    /**
     * Retrieves a page of a provider's accounts, with a low-level request cursor.
     *
     * @param provider    the provider, required
     * @param label       the provider configuration label, required
     * @param accountType filter by account type (optional)
     * @param sortOrder   sort order for results (optional, "ASC" or "DESC")
     * @param cursor      the request cursor, null for the first page with the default size
     * @return the accounts and their page
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if provider or label is empty
     */
    public FiatProviderAccountResult getFiatProviderAccounts(final String provider, final String label,
                                                              final String accountType, final String sortOrder,
                                                              final ApiRequestCursor cursor)
            throws ApiException {
        checkArgument(!Strings.isNullOrEmpty(provider), "provider cannot be null or empty");
        checkArgument(!Strings.isNullOrEmpty(label), "label cannot be null or empty");
        final CursorRequest page = CursorRequest.of(cursor);

        try {
            TgvalidatordGetFiatProviderAccountsReply reply = fiatApi.fiatProviderServiceGetFiatProviderAccounts(
                    provider,
                    label,
                    sortOrder,
                    page.currentPage(),
                    page.pageRequest(),
                    page.pageSizeParam(),
                    accountType
            );
            return page.complete(mapper.fromAccountsReply(reply), PagedOperation.FIAT_PROVIDER_ACCOUNTS);
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a fiat provider counterparty account by ID.
     *
     * @param id the counterparty account ID
     * @return the fiat provider counterparty account
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if id is null or empty
     */
    public FiatProviderCounterpartyAccount getFiatProviderCounterpartyAccount(final String id) throws ApiException {
        checkArgument(!Strings.isNullOrEmpty(id), "id cannot be null or empty");

        try {
            TgvalidatordGetFiatProviderCounterpartyAccountReply reply =
                    fiatApi.fiatProviderServiceGetFiatProviderCounterpartyAccount(id);
            return mapper.fromCounterpartyAccountDTO(reply.getResult());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a page of a provider's counterparty accounts.
     *
     * @param provider       the provider, required
     * @param label          the provider configuration label, required
     * @param counterpartyId filter by counterparty ID (optional)
     * @param sortOrder      sort order for results (optional, "ASC" or "DESC")
     * @param pageSize       the page size, null or 0 for the default
     * @param cursor         a previous page's {@code getPage().getNextCursor()}, null for the first page
     * @return the counterparty accounts and their page
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if provider or label is empty, or the page size is out of range
     */
    public FiatProviderCounterpartyAccountResult getFiatProviderCounterpartyAccounts(
            final String provider, final String label, final String counterpartyId,
            final String sortOrder, final Integer pageSize, final String cursor)
            throws ApiException {
        return getFiatProviderCounterpartyAccounts(provider, label, counterpartyId, sortOrder,
                Pagination.page(pageSize, cursor));
    }

    /**
     * Retrieves a page of a provider's counterparty accounts, with a low-level request cursor.
     *
     * @param provider       the provider, required
     * @param label          the provider configuration label, required
     * @param counterpartyId filter by counterparty ID (optional)
     * @param sortOrder      sort order for results (optional, "ASC" or "DESC")
     * @param cursor         the request cursor, null for the first page with the default size
     * @return the counterparty accounts and their page
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if provider or label is empty
     */
    public FiatProviderCounterpartyAccountResult getFiatProviderCounterpartyAccounts(
            final String provider, final String label, final String counterpartyId,
            final String sortOrder, final ApiRequestCursor cursor)
            throws ApiException {
        checkArgument(!Strings.isNullOrEmpty(provider), "provider cannot be null or empty");
        checkArgument(!Strings.isNullOrEmpty(label), "label cannot be null or empty");
        final CursorRequest page = CursorRequest.of(cursor);

        try {
            TgvalidatordGetFiatProviderCounterpartyAccountsReply reply =
                    fiatApi.fiatProviderServiceGetFiatProviderCounterpartyAccounts(
                            provider,
                            label,
                            counterpartyId,
                            sortOrder,
                            page.currentPage(),
                            page.pageRequest(),
                            page.pageSizeParam()
                    );
            return page.complete(mapper.fromCounterpartyAccountsReply(reply),
                    PagedOperation.FIAT_PROVIDER_COUNTERPARTY_ACCOUNTS);
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a fiat provider operation by ID.
     *
     * @param id the operation ID
     * @return the fiat provider operation
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if id is null or empty
     */
    public FiatProviderOperation getFiatProviderOperation(final String id) throws ApiException {
        checkArgument(!Strings.isNullOrEmpty(id), "id cannot be null or empty");

        try {
            TgvalidatordGetFiatProviderOperationReply reply =
                    fiatApi.fiatProviderServiceGetFiatProviderOperation(id);
            return mapper.fromOperationDTO(reply.getResult());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a page of fiat provider operations.
     *
     * @param provider  filter by provider (optional)
     * @param label     filter by label (optional)
     * @param sortOrder sort order for results (optional, "ASC" or "DESC")
     * @param pageSize  the page size, null or 0 for the default
     * @param cursor    a previous page's {@code getPage().getNextCursor()}, null for the first page
     * @return the operations and their page
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if the page size is out of range
     */
    public FiatProviderOperationResult getFiatProviderOperations(final String provider, final String label,
                                                                  final String sortOrder, final Integer pageSize,
                                                                  final String cursor)
            throws ApiException {
        return getFiatProviderOperations(provider, label, sortOrder, Pagination.page(pageSize, cursor));
    }

    /**
     * Retrieves a page of fiat provider operations, with a low-level request cursor.
     *
     * @param provider  filter by provider (optional)
     * @param label     filter by label (optional)
     * @param sortOrder sort order for results (optional, "ASC" or "DESC")
     * @param cursor    the request cursor, null for the first page with the default size
     * @return the operations and their page
     * @throws ApiException if the API call fails
     */
    public FiatProviderOperationResult getFiatProviderOperations(final String provider, final String label,
                                                                  final String sortOrder,
                                                                  final ApiRequestCursor cursor)
            throws ApiException {
        final CursorRequest page = CursorRequest.of(cursor);

        try {
            TgvalidatordGetFiatProviderOperationsReply reply = fiatApi.fiatProviderServiceGetFiatProviderOperations(
                    provider,
                    label,
                    sortOrder,
                    page.currentPage(),
                    page.pageRequest(),
                    page.pageSizeParam()
            );
            return page.complete(mapper.fromOperationsReply(reply), PagedOperation.FIAT_PROVIDER_OPERATIONS);
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a page of the entities registered with fiat providers.
     *
     * @param provider  filter by provider (optional)
     * @param label     filter by label (optional)
     * @param sortOrder sort order for results (optional, "ASC" or "DESC")
     * @param pageSize  the page size, null or 0 for the default
     * @param cursor    a previous page's {@code getPage().getNextCursor()}, null for the first page
     * @return the entities and their page
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if the page size is out of range
     */
    public FiatProviderEntityResult listFiatProviderEntities(final String provider, final String label,
                                                             final String sortOrder, final Integer pageSize,
                                                             final String cursor) throws ApiException {
        final CursorRequest page = CursorRequest.of(Pagination.page(pageSize, cursor));

        try {
            TgvalidatordGetFiatProviderEntitiesReply reply = fiatApi.fiatProviderServiceGetFiatProviderEntities(
                    provider,
                    label,
                    sortOrder,
                    page.currentPage(),
                    page.pageRequest(),
                    page.pageSizeParam()
            );
            FiatProviderEntityResult result = new FiatProviderEntityResult();
            result.setEntities(reply.getResult() == null
                    ? Collections.emptyList() : mapper.fromEntityDTOList(reply.getResult()));
            return page.complete(result, PagedOperation.FIAT_PROVIDER_ENTITIES, reply.getCursor(), null);
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }
}

File: taurus-protect-sdk-typescript/src/services/fiat-service.ts
/**
 * Fiat service for Taurus-PROTECT SDK.
 *
 * Provides methods for managing fiat provider accounts, counterparty accounts,
 * and operations in the Taurus-PROTECT system.
 */

import { ValidationError } from '../errors';
import type { FiatApi } from '../internal/openapi/apis/FiatApi';
import {
  fiatProvidersFromDto,
  fiatProviderAccountFromDto,
  fiatProviderAccountsFromDto,
  fiatProviderCounterpartyAccountFromDto,
  fiatProviderCounterpartyAccountsFromDto,
  fiatProviderEntitiesFromDto,
  fiatProviderOperationFromDto,
  fiatProviderOperationsFromDto,
} from '../mappers/fiat';
import type {
  FiatProvider,
  FiatProviderAccount,
  FiatProviderAccountResult,
  FiatProviderCounterpartyAccount,
  FiatProviderCounterpartyAccountResult,
  FiatProviderOperation,
  FiatProviderOperationResult,
  ListFiatProviderAccountsOptions,
  ListFiatProviderCounterpartyAccountsOptions,
  ListFiatProviderEntitiesOptions,
  ListFiatProviderEntitiesResult,
  ListFiatProviderOperationsOptions,
} from '../models/fiat';
import { buildCursorPage, cursorRequest } from '../models/pagination';
import { BaseService } from './base';
import { cursorQuery } from './paging';

/**
 * Service for managing fiat provider operations in the Taurus-PROTECT system.
 *
 * This service provides access to fiat provider accounts, counterparty accounts,
 * and operations for fiat currency management.
 *
 * @example
 * ```typescript
 * // List all fiat providers
 * const providers = await fiatService.getFiatProviders();
 *
 * // Get a specific account
 * const account = await fiatService.getFiatProviderAccount('account-123');
 *
 * // List accounts with filtering
 * const result = await fiatService.getFiatProviderAccounts({
 *   provider: 'circle',
 *   label: 'main',
 * });
 * ```
 */
export class FiatService extends BaseService {
  private readonly fiatApi: FiatApi;

  /**
   * Creates a new FiatService instance.
   *
   * @param fiatApi - The FiatApi instance from the OpenAPI client
   */
  constructor(fiatApi: FiatApi) {
    super();
    this.fiatApi = fiatApi;
  }

  /**
   * Retrieves all configured fiat providers.
   *
   * @returns Array of fiat providers
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const providers = await fiatService.getFiatProviders();
   * for (const provider of providers) {
   *   console.log(`${provider.provider}: ${provider.baseCurrencyValuation}`);
   * }
   * ```
   */
  async getFiatProviders(): Promise<FiatProvider[]> {
    return this.execute(async () => {
      const response = await this.fiatApi.fiatProviderServiceGetFiatProviders();
      return fiatProvidersFromDto(response.fiatProviders);
    });
  }

  /**
   * Retrieves a fiat provider account by ID.
   *
   * @param id - The account ID
   * @returns The fiat provider account
   * @throws {@link ValidationError} If id is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const account = await fiatService.getFiatProviderAccount('account-123');
   * console.log(`Balance: ${account.totalBalance} ${account.currencyId}`);
   * ```
   */
  async getFiatProviderAccount(id: string): Promise<FiatProviderAccount> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      const response =
        await this.fiatApi.fiatProviderServiceGetFiatProviderAccount({ id });
      const result = response.result;
      const account = fiatProviderAccountFromDto(result);
      if (!account) {
        throw new ValidationError(`Fiat provider account '${id}' not found`);
      }
      return account;
    });
  }

  /**
   * Retrieves a page of fiat provider accounts.
   *
   * @param options - Provider and label (required), filters, `pageSize` (1-100,
   *   default 20) and `cursor`
   * @returns The page of accounts and its cursor pagination
   * @throws {@link ValidationError} If required options are missing or paging options
   *   are invalid
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const result = await fiatService.getFiatProviderAccounts({
   *   provider: 'circle',
   *   label: 'main',
   *   accountType: 'wallet',
   *   sortOrder: 'ASC',
   * });
   * for (const account of result.accounts) {
   *   console.log(`${account.accountName}: ${account.totalBalance}`);
   * }
   * ```
   */
  async getFiatProviderAccounts(
    options: ListFiatProviderAccountsOptions
  ): Promise<FiatProviderAccountResult> {
    if (!options.provider || options.provider.trim() === '') {
      throw new ValidationError('provider is required');
    }
    if (!options.label || options.label.trim() === '') {
      throw new ValidationError('label is required');
    }

    const page = cursorRequest(options);

    return this.execute(async () => {
      const response =
        await this.fiatApi.fiatProviderServiceGetFiatProviderAccounts({
          provider: options.provider,
          label: options.label,
          accountType: options.accountType,
          sortOrder: options.sortOrder,
          ...cursorQuery(page),
        });
      return {
        accounts: fiatProviderAccountsFromDto(response.result),
        pagination: buildCursorPage(page.pageSize, response.cursor),
      };
    });
  }

  /**
   * Retrieves a fiat provider counterparty account by ID.
   *
   * @param id - The counterparty account ID
   * @returns The fiat provider counterparty account
   * @throws {@link ValidationError} If id is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const account = await fiatService.getFiatProviderCounterpartyAccount('cp-123');
   * console.log(`Counterparty: ${account.counterpartyName}`);
   * ```
   */
  async getFiatProviderCounterpartyAccount(
    id: string
  ): Promise<FiatProviderCounterpartyAccount> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      const response =
        await this.fiatApi.fiatProviderServiceGetFiatProviderCounterpartyAccount(
          { id }
        );
      const result = response.result;
      const account = fiatProviderCounterpartyAccountFromDto(result);
      if (!account) {
        throw new ValidationError(
          `Fiat provider counterparty account '${id}' not found`
        );
      }
      return account;
    });
  }

  /**
   * Retrieves a page of fiat provider counterparty accounts.
   *
   * @param options - Provider and label (required), filters, `pageSize` (1-100,
   *   default 20) and `cursor`
   * @returns The page of counterparty accounts and its cursor pagination
   * @throws {@link ValidationError} If required options are missing or paging options
   *   are invalid
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const result = await fiatService.getFiatProviderCounterpartyAccounts({
   *   provider: 'cubnet',
   *   label: 'main',
   *   counterpartyId: 'counterparty-123',
   * });
   * for (const account of result.accounts) {
   *   console.log(`${account.counterpartyName}: ${account.accountType}`);
   * }
   * ```
   */
  async getFiatProviderCounterpartyAccounts(
    options: ListFiatProviderCounterpartyAccountsOptions
  ): Promise<FiatProviderCounterpartyAccountResult> {
    if (!options.provider || options.provider.trim() === '') {
      throw new ValidationError('provider is required');
    }
    if (!options.label || options.label.trim() === '') {
      throw new ValidationError('label is required');
    }

    const page = cursorRequest(options);

    return this.execute(async () => {
      const response =
        await this.fiatApi.fiatProviderServiceGetFiatProviderCounterpartyAccounts({
          provider: options.provider,
          label: options.label,
          counterpartyID: options.counterpartyId,
          sortOrder: options.sortOrder,
          ...cursorQuery(page),
        });
      return {
        accounts: fiatProviderCounterpartyAccountsFromDto(response.result),
        pagination: buildCursorPage(page.pageSize, response.cursor),
      };
    });
  }

  /**
   * Retrieves a fiat provider operation by ID.
   *
   * @param id - The operation ID
   * @returns The fiat provider operation
   * @throws {@link ValidationError} If id is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const operation = await fiatService.getFiatProviderOperation('op-123');
   * console.log(`Status: ${operation.status}, Amount: ${operation.amount}`);
   * ```
   */
  async getFiatProviderOperation(id: string): Promise<FiatProviderOperation> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      const response =
        await this.fiatApi.fiatProviderServiceGetFiatProviderOperation({ id });
      const result = response.result;
      const operation = fiatProviderOperationFromDto(result);
      if (!operation) {
        throw new ValidationError(`Fiat provider operation '${id}' not found`);
      }
      return operation;
    });
  }

  /**
   * Retrieves a page of fiat provider operations.
   *
   * @param options - Filters, `pageSize` (1-100, default 20) and `cursor`
   * @returns The page of operations and its cursor pagination
   * @throws {@link ValidationError} If the paging options are invalid
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List all operations
   * const result = await fiatService.getFiatProviderOperations();
   *
   * // List operations with filtering
   * const filtered = await fiatService.getFiatProviderOperations({
   *   provider: 'circle',
   *   label: 'main',
   *   sortOrder: 'DESC',
   * });
   * for (const op of filtered.operations) {
   *   console.log(`${op.operationType}: ${op.amount} (${op.status})`);
   * }
   * ```
   */
  async getFiatProviderOperations(
    options?: ListFiatProviderOperationsOptions
  ): Promise<FiatProviderOperationResult> {
    const page = cursorRequest(options);

    return this.execute(async () => {
      const response =
        await this.fiatApi.fiatProviderServiceGetFiatProviderOperations({
          provider: options?.provider,
          label: options?.label,
          sortOrder: options?.sortOrder,
          ...cursorQuery(page),
        });
      return {
        operations: fiatProviderOperationsFromDto(response.result),
        pagination: buildCursorPage(page.pageSize, response.cursor),
      };
    });
  }

  /**
   * Lists a page of fiat provider entities.
   *
   * @param options - Filters, `pageSize` (1-100, default 20) and `cursor`
   * @returns The page of entities and its cursor pagination
   * @throws {@link ValidationError} If the page size is out of bounds
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * let cursor: string | undefined;
   * do {
   *   const page = await fiatService.listFiatProviderEntities({ provider: 'bank', cursor });
   *   page.items.forEach((e) => console.log(e.name));
   *   cursor = page.pagination.hasMore ? page.pagination.nextCursor : undefined;
   * } while (cursor);
   * ```
   */
  async listFiatProviderEntities(
    options?: ListFiatProviderEntitiesOptions
  ): Promise<ListFiatProviderEntitiesResult> {
    const page = cursorRequest(options);

    return this.execute(async () => {
      const response =
        await this.fiatApi.fiatProviderServiceGetFiatProviderEntities({
          provider: options?.provider,
          label: options?.label,
          sortOrder: options?.sortOrder,
          ...cursorQuery(page),
        });
      return {
        items: fiatProviderEntitiesFromDto(response.result),
        pagination: buildCursorPage(page.pageSize, response.cursor),
      };
    });
  }
}
